idp-binding-service/src/main/java/io/mosip/idp/binding/services/WalletBindingServiceImpl.java

```diff
@@ -0,0 +1,213 @@
+
+	@Autowired
+	private CacheUtilService cacheUtilService;
+
+	@Autowired
+	private AuthenticationWrapper authenticationWrapper;
+
+	@Autowired
+	private PublicKeyRegistryRepository publicKeyRegistryRepository;
+
+	@Value("${mosip.idp.binding.auth-partner-id}")
+	private String authPartnerId;
+
+	@Value("${mosip.idp.binding.auth-license-key}")
+	private String apiKey;
+
+	@Value("${mosip.idp.binding.public-key-expire-days}")
+	private int expireDays;
+
+	@Value("${mosip.idp.binding.issuer-id}")
+	private String issuerId;
+
+	@Value("${mosip.idp.binding.salt-length}")
+	private int saltLength;
+
+	@Override
+	public void sendBindingOtp() throws IdPException {
+		// TODO Auto-generated method stub
+
+	}
+
+	@Override
+	public WalletBindingResponse bindWallet(WalletBindingRequest walletBindingRequest) throws IdPException {
+		BindingTransaction bindingTransaction = cacheUtilService
+				.getTransaction(walletBindingRequest.getTransactionId());
+		if (bindingTransaction == null)
+			throw new InvalidTransactionException();
+
+		if (!bindingTransaction.getIndividualId().equals(walletBindingRequest.getIndividualId()))
+			throw new IdPException(INVALID_INDIVIDUAL_ID);
+
+		if (!bindingTransaction.getAuthChallengeType()
+				.equals(walletBindingRequest.getChallengeList().get(0).getAuthFactorType()))
+			throw new IdPException(INVALID_AUTH_CHALLENGE);
+
+		log.info("Wallet Binding Request validated and sent for authentication");
+
+		KycAuthResult kycAuthResult = authenticateIndividual(bindingTransaction, walletBindingRequest);
+
+		log.info("Wallet Binding Request authentication is successful");
+
+		PublicKeyRegistry publicKeyRegistry = storeData(walletBindingRequest,
+				kycAuthResult.getPartnerSpecificUserToken());
+
+		WalletBindingResponse walletBindingResponse = new WalletBindingResponse();
+		walletBindingResponse.setTransactionId(walletBindingRequest.getTransactionId());
+		walletBindingResponse.setExpireDateTime(
+				publicKeyRegistry.getExpiredtimes().format(DateTimeFormatter.ofPattern(UTC_DATETIME_PATTERN)));
+		// TODO need to call Keymanager publickey-JWE encryption
+		walletBindingResponse.setEncryptedWalletBindingId(
+				getEncryptedWalletBindingId(walletBindingRequest, publicKeyRegistry,
+						publicKeyRegistry.getWalletBindingId()));
+
+		return walletBindingResponse;
+
+	}
+
+	@Override
+	public void validateBinding() throws IdPException {
+		// TODO Auto-generated method stub
+
+	}
+
+	private PublicKeyRegistry storeData(WalletBindingRequest walletBindingRequest,
+			String partnerSpecificUserToken) throws IdPException {
+		PublicKeyRegistry publicKeyRegistry;
+		String publicKey = IdentityProviderUtil.getJWKString(walletBindingRequest.getPublicKey());
+		String publicKeyHash=IdentityProviderUtil.generateB64EncodedHash(ALGO_SHA_256,publicKey );
+		LocalDateTime expiredtimes = calculateExpiresdtimes();
+		Optional<PublicKeyRegistry> optionalPublicKeyRegistry = publicKeyRegistryRepository
```

Comment:
This could return more than 1 row. Instead of querying based on psu token, can we query it based on the individualId and insert one if not present?
if an entry is found or not found for the input individual id we should always run "update  public_key_registry set public_key=?, public_key_hash=?,expire_dtimes=? where psu_token=?"

Reply:
Fixed